gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/WsUserAgentHandler.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.server.handler;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.websocketx.WebSocketServerProtocolHandler;
+import io.netty.util.AttributeKey;
+import org.apache.tinkerpop.gremlin.driver.util.UserAgent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Channel handler which extracts a user agent header from a web socket handshake if present
+ * then logs the user agent and stores it as a channel attribute for future reference.
+ */
+public class WsUserAgentHandler extends ChannelInboundHandlerAdapter {
```

Comment:
is `info` the right log level here? seems we would trigger this for each connection made by a driver. maybe that would flood the logs with too much stuff? maybe we `debug` logging is better and some sort of aggregation of agent types reported as metrics (which by default only periodically writes to the logs)? thoughts?

Reply:
I agree, best to keep logs clearer and aggregated user agents will be more useful too.

Reply:
you went with `UserAgent` for the metric name, but please double check to see if that is consistent with patterns already established for naming metrics in Gremlin Server. i think naming prefixes are all lower case with a hyphen between words if there are more than one.  also, since you added a metric, please document it [here](https://tinkerpop.apache.org/docs/current/reference/#metrics).

Reply:
I've updated the name to follow the existing pattern. Unfortunately because the user agent itself is being used as the final part of the metric name, that portion of it will not match the existing format.

Reply:
> that portion of it will not match the existing format.

I believe that is fine as I think other metrics work that way as well. as long as the static prefix part of the metric is named according to standard the dynamic right-hand-side can do as it wishes.